packages/hash/shared/src/prosemirror.ts

```diff
@@ -14,120 +14,131 @@ export type EntityNode = NodeWithAttrs<{
-       * plain text
-       *
-       * @todo figure out to do this here
-       * @see createEditorView
-       */
-      hardBreak: {
-        inline: true,
-        group: "inline",
-        selectable: false,
-        parseDOM: [{ tag: "br" }],
-        toDOM() {
-          return ["br"];
-        },
-      },
-      mention: {
-        inline: true,
-        group: "inline",
-        atom: true,
-        attrs: { mentionType: { default: null }, entityId: { default: null } },
-        toDOM: (node) => {
-          const { mentionType, entityId } = node.attrs;
-          return [
-            "span",
-            {
-              "data-hash-type": "mention",
-              "data-mention-type": mentionType,
-              "data-entity-id": entityId,
-            },
-          ] as const;
-        },
-        parseDOM: [
-          {
-            tag: 'span[data-hash-type="mention"]',
-            getAttrs(dom) {
-              return {
-                mentionType: (dom as Element).getAttribute("data-mention-type"),
-                entityId: (dom as Element).getAttribute("data-entity-id"),
-              };
-            },
-          },
-        ],
+    },
+  ],
+};
+
+export const loadingNode: NodeSpec = {
+  /**
+   * As we don't have any component nodes defined by default, we need a
+   * placeholder, otherwise Prosemirror will crash when trying to
+   * interpret the content expressions in other nodes. However, as soon
+   * as we have defined a different component node, we remove the blank
+   * node from the componentNode group, which ensures that when
+   * Prosemirror attempts to instantiate a componentNode it uses that
+   * node instead of the blank one
+   *
+   * @see import("./ProsemirrorManager.ts").ProsemirrorManager#prepareToDisableBlankDefaultComponentNode
+   */
+  group: componentNodeGroupName,
+  toDOM: () => ["div", 0] as const,
+};
+
+export const blockNode: NodeSpec = {
+  content: "entity",
+  /**
+   * These properties are necessary for copy and paste (which is
+   * necessary for drag and drop)
+   *
+   * @note – the actual rendering in the DOM is taken over by the NodeView
+   *         so check `BlockView` and `ComponentView` for how this will
+   *         actually appear
+   */
+  toDOM: () => {
+    return [
+      "div",
+      {
+        "data-hash-type": "block",
       },
+      0,
+    ] as const;
+  },
+  parseDOM: [
+    {
+      tag: 'div[data-hash-type="block"]',
     },
+  ],
+};
+
+export const entityNode: NodeSpec = {
+  content: "componentNode | entity",
+  attrs: {
+    draftId: { default: null },
+  },
+  toDOM: () => {
+    return ["div", { "data-hash-type": "entity" }, 0] as const;
+  },
+  parseDOM: [
+    {
+      tag: 'div[data-hash-type="entity"]',
+    },
+  ],
+};
+
+export const getPageEditorNodes = (): NodeSpecs => ({
+  doc: {
+    content: "((componentNode|block)+)|loading",
+  },
+  text: textNode,
+  hardBreak: hardBreakNode,
+  mention: mentionNode,
+  loading: { ...loadingNode },
```

Comment:
What's the spreading here necessary for?

Reply:
`mutateSchema` was mutating the `loading` node, causing the app to crash whenever we loaded a new page (because the loading node wouldn't have the `componentNode` group). I didn't understand this at the time and was the quick fix I found. I meant to investigating it and finding a more elegant solution before sending it for review but it slipped through. Thanks for catching it!